### PR TITLE
Implement Batch Group Arithmetic utilising Montgomery's trick & w-NAF precomputation

### PR DESCRIPTION
**Pull request by jon-chuang**
Single-core speedup 1.75-1.85x. Multicore speedup: 1.5x.

### CODE REVIEW

File: algebra-core/field-assembly/src/lib.rs [hunk 1]
```diff
@@ -102,3 +79,26 @@ fn generate_matches(num_limbs: usize, is_mul: bool) -> String {
     ctx.end(num_limbs);
     ctx.get_string()
 }
+
+pub fn generate_macro_string(num_limbs: usize) -> std::string::String {
```
Comment (kobigurk) [inline algebra-core/field-assembly/src/lib.rs:83]: Has anything changed here?

File: algebra-core/src/curves/mod.rs [hunk 2]
```diff
@@ -308,3 +310,218 @@ where
...
+
+pub trait BatchGroupArithmetic
+where
+    Self: Sized + Clone + Copy + Zero + Neg<Output = Self>,
+{
+    // This function consumes the scalars
+    // We can make this more generic in the future to use other than u16.
+
+    // TODO: Generalise to A != 0
+    // Computes [-p, p, -3p, 3p, ..., -2^wp, 2^wp]
```
Comment (kobigurk) [inline algebra-core/src/curves/mod.rs:322]: Better to use Rust-style comments here: `///` instead of `//`.

File: algebra-core/src/curves/mod.rs [hunk 3]
```diff
@@ -308,3 +310,218 @@ where
         Fq = <Self::E1 as PairingEngine>::Fr,
     >;
 }
+
+pub trait BatchGroupArithmetic
+where
+    Self: Sized + Clone + Copy + Zero + Neg<Output = Self>,
+{
+    // This function consumes the scalars
+    // We can make this more generic in the future to use other than u16.
```
Comment (kobigurk) [inline algebra-core/src/curves/mod.rs:319]: Let's move this into an issue rather than a comment.

File: algebra-core/src/curves/mod.rs [hunk 4]
```diff
@@ -308,3 +310,218 @@ where
...
+
+    // TODO: Generalise to A != 0
```
Comment (kobigurk) [inline algebra-core/src/curves/mod.rs:321]: Let's move this into an issue rather than a comment.

File: algebra-core/src/curves/mod.rs [hunk 5]
```diff
@@ -308,3 +310,218 @@ where
...
+
+            for (table, p) in tables.iter_mut().zip(&tmp) {
+                table.push(p.clone());
+            }
+        }
+        tables
+    }
+
+    // This function mutates the scalars in place
+    // We can make this more generic in the future to use other than u16.
```
Comment (kobigurk) [inline algebra-core/src/curves/mod.rs:351]: Better to use Rust-style comments here: /// instead of //.

File: algebra-core/src/curves/mod.rs [hunk 6]
```diff
@@ -308,3 +310,218 @@ where
...
+            all_none = opcode_row.iter().all(|x| x.is_none());
+            if !all_none {
+                op_code_vectorised.push(opcode_row);
+            }
+        }
+        op_code_vectorised
+    }
+
+    // This function consumes the second op as it mutates it in place
+    // to prevent memory allocation
```
Comment (kobigurk) [inline algebra-core/src/curves/mod.rs:399]: Better to use Rust-style comments here: /// instead of //.

File: algebra-core/src/serialize/mod.rs [hunk 7]
```diff
@@ -407,9 +407,9 @@ macro_rules! impl_sw_curve_serializer {
                     CanonicalDeserializeWithFlags::deserialize_with_flags(reader)?;
 
                 let p = GroupAffine::<P>::new(x, y, flags.is_infinity());
-                if !p.is_in_correct_subgroup_assuming_on_curve() {
-                    return Err(crate::serialize::SerializationError::InvalidData);
-                }
+                // if !p.is_in_correct_subgroup_assuming_on_curve() {
```
Comment (kobigurk) [inline algebra-core/src/serialize/mod.rs:410]: Let's revert this check back and move to use `deserialize_unchecked`.

File: algebra-core/src/curves/mod.rs [hunk 8]
```diff
@@ -308,3 +310,218 @@ where
...
+            let index_double = opcode_row
+                .iter()
+                .enumerate()
+                .filter(|x| x.1.is_some())
+                .map(|x| x.0)
+                .collect();
+
+            Self::batch_double_in_place(&mut bases, index_double);
+
+            // Copying to this vector might be really stupid...
```
Comment (kobigurk) [inline algebra-core/src/curves/mod.rs:428]: Why?

File: algebra-core/src/curves/mod.rs [hunk 9]
```diff
@@ -308,3 +310,218 @@ where
...
+    fn batch_add_in_place(&mut self, other: &mut Self, index: Vec<(usize, usize)>) {
+        G::batch_add_in_place(self, other, index);
+    }
+
+    fn batch_scalar_mul_in_place<BigInt: BigInteger>(&mut self, scalars: &mut [BigInt], w: usize) {
+        G::batch_scalar_mul_in_place(self, scalars, w);
+    }
+}
+
+trait GLV: AffineCurve {
```
Comment (kobigurk) [inline algebra-core/src/curves/mod.rs:498]: Since the endomorphism is unimplemented, how about we move it to another PR?

File: algebra-core/src/curves/models/short_weierstrass_affine.rs [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+                self.mul(P::COFACTOR_INV).into()
+            }
+        }
+
+        impl<P: Parameters> BatchGroupArithmetic for GroupAffine<P> {
+            // This implementation of batch group ops takes particular
+            // care to make most use of points fetched from memory to prevent reallocations
+            // It is adapted from Aztec's code.
+
+            // https://github.com/AztecProtocol/barretenberg/blob/standardplonkjson/barretenberg/src/
```
Comment (kobigurk) [inline algebra-core/src/curves/models/short_weierstrass_affine.rs:74]: Let's add the commit as well.

File: algebra-core/src/curves/models/twisted_edwards_extended.rs [hunk 11]
```diff
@@ -165,6 +165,64 @@ impl<P: Parameters> AffineCurve for GroupAffine<P> {
...
+        );
+    }
+
+    // Total cost: 14 mul. Projective formulas:
+    fn batch_add_in_place(bases: &mut [Self], other: &mut [Self], index: Vec<(usize, usize)>) {
+        let mut inversion_tmp = P::BaseField::one();
+        // We run two loops over the data separated by an inversion
+        for (idx, idy) in index.iter() {
+            let (mut a, mut b) = (&mut bases[*idx], &mut other[*idy]);
+            if a.is_zero() || b.is_zero() {
```
Comment (kobigurk) [inline algebra-core/src/curves/models/twisted_edwards_extended.rs:185]: Any chance it would be better to filter these out?
Maybe not because then we do 2 passes.

File: algebra-core/src/curves/models/twisted_edwards_extended.rs [hunk 12]
```diff
@@ -165,6 +165,64 @@ impl<P: Parameters> AffineCurve for GroupAffine<P> {
...
+    // to prevent memory allocation
+    fn batch_double_in_place(bases: &mut [Self], index: Vec<usize>) {
+        Self::batch_add_in_place(
+            bases,
+            &mut bases.to_vec()[..],
+            index.iter().map(|&x| (x, x)).collect(),
+        );
+    }
+
+    // Total cost: 14 mul. Projective formulas:
```
Comment (kobigurk) [inline algebra-core/src/curves/models/twisted_edwards_extended.rs:179]: Can you link the projective formulas you've used?

File: algebra-core/src/curves/models/short_weierstrass_affine.rs [hunk 13]
```diff
@@ -0,0 +1,399 @@
...
+                        if let Some((idp_1, idp_2)) = prefetch_iter.next() {
+                            prefetch::<Self>(&mut bases[*idp_1]);
+                            prefetch::<Self>(&mut other[*idp_2]);
+                        }
+                    }
+                    let (mut a, mut b) = (&mut bases[*idx], &mut other[*idy]);
+                    if a.is_zero() || b.is_zero() {
+                        continue;
+                    } else if a.x == b.x {
+                        half = match half {
```
Comment (kobigurk) [inline algebra-core/src/curves/models/short_weierstrass_affine.rs:180]: Why don't we initialize `half` before?

File: algebra-core/src/curves/models/short_weierstrass_affine.rs [hunk 14]
```diff
@@ -0,0 +1,399 @@
...
+                        a.x -= &b.x; // denominator = x1 - x2
+                        a.y -= &b.y; // numerator = y1 - y2
+                        a.y *= &inversion_tmp; // (y1 - y2)*tmp
+                        inversion_tmp *= &a.x // update tmp
+                    }
+                }
+
+                inversion_tmp = inversion_tmp.inverse().unwrap(); // this is always in Fp*
+
+                #[cfg(feature = "prefetch")]
```
Comment (kobigurk) [inline algebra-core/src/curves/models/short_weierstrass_affine.rs:215]: You can possibly use `cfg_if` to make this cleaner.

File: algebra-core/src/curves/models/short_weierstrass_affine.rs [hunk 15]
```diff
@@ -0,0 +1,399 @@
+#[macro_export]
+macro_rules! specialise_affine_to_proj {
```
Comment (kobigurk) [inline algebra-core/src/curves/models/short_weierstrass_affine.rs:2]: Love this macro.

File: algebra-core/src/curves/mod.rs (hunk 4 above)
Comment (kobigurk) [reply algebra-core/src/curves/mod.rs:321]: Also, let's assert on it?

File: algebra-core/src/curves/mod.rs (hunk 2 above)
Comment (kobigurk) [reply algebra-core/src/curves/mod.rs:322]: Where does negation happen?

File: algebra-core/src/curves/mod.rs [hunk 16]
```diff
@@ -308,3 +310,218 @@ where
...
+                .enumerate()
+                .map(|(x, y)| (y, x))
+                .collect();
+
+            Self::batch_add_in_place(&mut bases, &mut add_ops[..], index_add);
+        }
+    }
+}
+
+// We make the syntax cleaner by defining corresponding trait and impl for [G]
```
Comment (kobigurk) [inline algebra-core/src/curves/mod.rs:457]: Rust-style comment that is less personal :)

**Review by kobigurk — CHANGES_REQUESTED** [review]
Looks amazing. A few comments.

File: algebra-core/field-assembly/src/lib.rs (hunk 1 above)
Comment (jon-chuang) [reply algebra-core/field-assembly/src/lib.rs:83]: I just reordered it

File: algebra-core/src/curves/mod.rs [hunk 17]
```diff
@@ -308,3 +310,218 @@ where
...
+            for (table, p) in tables.iter_mut().zip(&tmp) {
+                table.push(p.clone());
+            }
+        }
+        tables
+    }
+
+    // This function mutates the scalars in place
+    // We can make this more generic in the future to use other than u16.
+    fn batch_wnaf_opcode_recoding<BigInt: BigInteger + AsRef<[u64]>>(
```
Comment (Pratyush) [inline algebra-core/src/curves/mod.rs:352]: `BigInteger` already requires `AsRef<[u64]>`, so you don't need to have another impl

```suggestion
    fn batch_wnaf_opcode_recoding<BigInt: BigInteger>(
```

File: algebra-core/src/curves/mod.rs [hunk 18]
```diff
@@ -308,3 +310,218 @@ where
...
+        for p in bases.iter_mut() {
+            *p = zero;
+        }
+
+        for opcode_row in opcode_vectorised.iter().rev() {
+            let index_double = opcode_row
+                .iter()
+                .enumerate()
+                .filter(|x| x.1.is_some())
+                .map(|x| x.0)
```
Comment (Pratyush) [inline algebra-core/src/curves/mod.rs:423]: Maybe replace this with [`filter_map`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.filter_map)?

File: algebra-core/src/curves/mod.rs [hunk 19]
```diff
@@ -308,3 +310,218 @@ where
...
+pub trait BatchGroupArithmetic
+where
+    Self: Sized + Clone + Copy + Zero + Neg<Output = Self>,
+{
+    // This function consumes the scalars
+    // We can make this more generic in the future to use other than u16.
+
+    // TODO: Generalise to A != 0
+    // Computes [-p, p, -3p, 3p, ..., -2^wp, 2^wp]
+    fn batch_wnaf_tables(bases: &[Self], w: usize) -> Vec<Vec<Self>> {
```
Comment (Pratyush) [inline algebra-core/src/curves/mod.rs:323]: Is `w` the window size? If so, it might make it slightly clearer to rename it to `window_size` or something similar (also elsewhere in the code).


Btw, is there a link to a description of the algorithm, something that isn't just Aztec's code?

File: algebra-core/src/curves/mod.rs (hunk 5 above)
Comment (Pratyush) [reply algebra-core/src/curves/mod.rs:351]: Also ideally the doc-comments should explain what the functions are doing (beyond describing whether an argument is consumed or mutated.

**Comment by kobigurk** [comment]
Can this be closed?
